### PR TITLE
[WIP] styles: Update style guide

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message: styles: Update style guide
Additional Description:

i have changed the focus of this as we will also need to update info for:

- yaml 
- general linting
- rst - including embedded in protos etc

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] Partial fix for #15403 , ref #15479 
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: STYLE.md [hunk 1]
```diff
@@ -246,8 +246,50 @@ such as the current time. Instead, mocks such as
...
+
+All python code is statically checked for correctness using pyflakes.
+
+## mypy (python typing)
+
+To be added.
+
+## linting
+
+All python code is linted using flake8
```
Comment (phlax) [inline STYLE.md:281]: add some instructions with/out bazel pointing to config file

**Comment by phlax** [comment]
re mypy the biggest issue that it introduces is py version.

we are currently pegged to `>=3.8` 

more recent versions of python - esp 3.9 have increasingly better lang-level support for some of the type hinting

older versions of python require use of the `typing` lib

File: STYLE.md [hunk 2]
```diff
@@ -246,8 +246,50 @@ such as the current time. Instead, mocks such as
...
+```python
+
+some_aggregate_condition = (
+    a_condition
+    and (a_sub_condition
+         or another_sub_condition))
+
+```
+
+In this example the line split occurs before the `and` and `or` which are placed at the
```
Comment (htuch) [inline STYLE.md:268]: Will this be lintable? I'd like this not to be a subject of code reviews :) FWIW this is the kind of thing that clang-format does well at.

File: STYLE.md [hunk 3]
```diff
@@ -246,8 +246,50 @@ such as the current time. Instead, mocks such as
 [`MockRandomGenerator`](test/mocks/runtime/mocks.h) and
 [`Mock*TimeSource`](test/mocks/common.h) should be used.
 
+# Python coding style
+
+## pep8
+
+All python code should meet the pep8/pycodestyle standard.
```
Comment (htuch) [inline STYLE.md:253]: Add links.

**Comment by htuch** [comment]
@phlax I like the idea of preferring f-strings (and other Python 3+ isms).

**Pull request closed without merging**